### DISCUSSION TITLE
Control the colors and line of the references as in hyperref

### DISCUSSION THREAD

**Discussion by Atreyu-94** [typst/typst#2585] [Ideas]
It would be great to be able to control the colors of the references/citations as you can control them in hyperref. This is very useful for creating documentation or educational content, for example:

|Lines|Colors|
|----------|--------|
|![imagen](https://github.com/typst/typst/assets/39834909/b1c82954-6ceb-4267-99f2-48021da34b4f) | ![imagen](https://github.com/typst/typst/assets/39834909/72db1ec2-c7af-47c6-aa2e-8017371bacbd) |

The way hyperref handles the layout is terrible but it is useful in the cases mentioned

**Comment by Dherse** [comment]
You can already do that using show and set rules on your references and links

**Reply by Atreyu-94** [reply]
Thank you! I haven't had time to review all the documentation, I thought this was not present. Do you have any example of use at hand? Thank you so much :)

**Reply by ghost** [reply]
I have added an example [here](https://github.com/typst/typst/discussions/2585#discussioncomment-7483688) to show that it is currently not possible with Typst.

**Comment by ghost** [comment]
I think you want the hyperref behaviour: have some color around the link when the pdf file is opened in a viewer, and have no color around the link if it is printed on a paper (the same with text color that disappears when printed).
If so, this is [not](https://github.com/typst/typst/discussions/2585#discussioncomment-7483688) currently possible. And I think there is also not an open issue currently.
But it is achievable at pdf level, so would be quite possible to implement this.

> The way hyperref handles the layout is terrible

Not sure what you mean exactly. I never had problems with it. If you mean the misplaced rectangles in the first picture, I never had those, it looks like a bug in your source code.

**Reply by Atreyu-94** [reply]
Thank you so much! This is exactly the behavior I was referring to. This afternoon, after class, I'm going to open an issue about it.

Regarding the layout, the examples I used were not mine, however many times in my own documents the layout had that strange behavior as well.

**Comment by LectronPusher** [comment]
Here's an example of a show rule with a few options: (edit: if you're ok with this style also appearing when printed)
```typ

// Show links with a box around them or filled with a color
#show link: this => {
  let show-type = "box" // "box" or "filled", see below
  let label-color = green
  let default-color = rgb("#ff66ff")
  
  if show-type == "box" {
    if type(this.dest) == label {
      // Make the box bound the entire text:
      set text(bottom-edge: "bounds", top-edge: "bounds")
      box(this, stroke: label-color + 1pt)
    } else {
      set text(bottom-edge: "bounds", top-edge: "bounds")
      box(this, stroke: default-color + 1pt)
    }
  } else if show-type == "filled" {
    if type(this.dest) == label {
      text(this, fill: label-color)
    } else {
      text(this, fill: default-color)
    }
  } else {
    this
  }
}

== Heading <some_heading>
https://example.com \
#link(<some_heading>)[Link to heading] \
#link("https://example.com") \
#link("https://example.com")[See example.com]
```
![image](https://github.com/typst/typst/assets/66051443/0c7bd670-a2a2-41c5-9c46-c7375fa9dc92)

**Reply by LectronPusher** [reply]
If you just want to add text or underlining to all links the same way, any of these one-liners should work:

```typ
// make links blue
#show link: set text(fill: rgb(0, 0, 255))

// underline links
#show link: underline

// surround links with a green box
#show link: box.with(stroke: green, inset: (bottom: 2pt))
```

**Reply by ghost** [reply]
As I have [written](https://github.com/typst/typst/discussions/2585#discussioncomment-7477560) in the comment above: this is not the hyperref behaviour, the boxes do not disappear when printed.

[url with typst.pdf](https://github.com/typst/typst/files/13263078/url.with.typst.pdf)

[url with hyperref.pdf](https://github.com/typst/typst/files/13263079/url.with.hyperref.pdf)

[printed_hyperref.pdf](https://github.com/typst/typst/files/13263100/printed_hyperref.pdf)

[printed_typst.pdf](https://github.com/typst/typst/files/13263101/printed_typst.pdf)

**Reply by LectronPusher** [reply]
Ah, apologies, I didn't read your comment closely enough. The original post wasn't clear that was what they wanted. Definitely would be a nice feature to have!

**Reply by Atreyu-94** [reply]
@LectronPusher Thank you very much too. I'm going to use your examples while the feature is implemented (if the developers decide)

**Reply by ziyuanding** [reply]
very nice example! I tried to make it work with `cite`, and the result is perfect.  But for `ref`, it's not what I expected. I hope there exists a way to only apply linking function and this colored box to the numbering part.
![ref_style](https://github.com/user-attachments/assets/66bb3a32-c6fe-4048-ae49-b85a480fd295)

**Reply by wrzian** [reply]
To do that you'll need to separate the styling of the supplement and the numbers.

Here's an example using a show rule on `ref`:

```typ
#let separate-supplement-style(supp, num) = {
  text(supp, fill: blue)
  [ ]
  box(num, stroke: 1pt + green, outset: (bottom:1.5pt, x:.5pt, y:.5pt))
}

#show ref: it => {
  let (element, target, supplement: supp) = it.fields()
  if element.func() != heading {
    return it
  }
  let head = element.fields()
  let supp = if supp == auto { head.supplement } else { supp }
  let num = context { // apply the heading's numbering style
    let head-count = counter(heading).at(target)
    numbering(head.numbering, ..head-count)
  }
  link(target, separate-supplement-style(supp, num))
}

#set heading(numbering: "1.")
= Head <head>
#lorem(10)
== Short Body <body>
#lorem(5)
== Long Tail <tail>
#lorem(50)

@head \
@body[hey its the body] \
@tail
```

![image](https://github.com/user-attachments/assets/4404c4e1-c0cf-45d9-95a9-b64abb2fe084)

**Reply by ziyuanding** [reply]
@wrzian Thanks! you just save my day!

``` typst
// I modified the code a bit to make it works along with cite and refs other than heading. note that equations don't have a numbering by default, so need to #set math.equation(numbering: "(1)") if anyone wants to reference an equation with a such colored box manner.
#show ref: it => {
  let (element, target, supplement: supp) = it.fields()
  // cite doesn't have element
  if element == none {
    return it
  }

  let non_cite_ref = element.fields()
  let supp = if supp == auto { non_cite_ref.supplement } else { supp }
  let num = context { // apply the heading's numbering style
    let head-count = counter(heading).at(target)
    numbering(non_cite_ref.numbering, ..head-count)
  }
  link(target, separate-supplement-style(supp, num))
}
```
